fix: reduce API test memory consumption from 8.26GB to 1.57GB (#8263)

highlyavailable opened the pull request:
# Memory Optimization Fix for API Tests - Issue #8263

API tests were consuming excessive memory (>40 GiB virtual, >6 GiB resident), causing slower test execution and reduced developer adoption.

## RCA

### Memory Profile Findings (Before Optimization)

- **Total Memory Allocation**: 8.26 GB
- **Ristretto Cache Dominance**: 6.7 GB (81% of total allocation)
  - `ristretto.newCmRow`: 4.5 GB (54.42%)
  - `ristretto.Bloom.Size`: 2.2 GB (27.21%)

### Root Cause

1. **Default Production Configuration in Tests**: Tests were using production-level cache sizes:
   - `committed.local_cache.size_bytes`: 1 GB (default)
   - `committed.sstable.memory.cache_size_bytes`: 400 MB (default)

2. **Ristretto Cache Overhead**: The ristretto cache allocates **10 million counters** regardless of actual cache capacity, creating massive memory overhead for test environments.

3. **Multiple Cache Instances**: Each test creates two pyramid filesystem caches:
   - **MetaRange FS**: ~50% of local cache allocation
   - **Range FS**: ~50% of local cache allocation

4. **Test Repetition**: 50+ controller tests each calling `setupHandler()` with fresh cache instances.

## Results

### Memory Allocation Improvement

- **Before**: 8.26 GB total allocation
- **After**: 1.57 GB total allocation
- **Reduction**: **81% memory reduction** (6.69 GB saved)

### Ristretto Cache Impact

- **Before**: 6.7 GB ristretto allocation (81% of total)
- **After**: 50 MB ristretto allocation (3.2% of total)
- **Ristretto Reduction**: **99.25% reduction** in ristretto overhead

### Testing

- Tests continue to pass with identical functionality
- Memory footprint now suitable for dev environments
- There shouldn't be any impact to prod configurations (only affects test environment)

## Verification

- Full controller test suite runs successfully
- Memory usage reduced from 8.26 GB to 1.57 GB
- No functional regressions detected
- Production configuration remains unchanged

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/treeverse/lakeFS?pullRequest=9097) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/treeverse/lakeFS?pullRequest=9097) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/treeverse/lakeFS?pullRequest=9097) it.</sub>

arielshaqed commented on pkg/api/serve_test.go:
```diff
@@ -113,6 +113,11 @@ func setupHandler(t testing.TB) (http.Handler, *dependencies) {
 
+	// Set minimal cache sizes for testing to reduce memory consumption
+	// Default production values are too large for test environments
+	viper.Set("committed.local_cache.size_bytes", 8*1024*1024)          // 8MB instead of 1GB
```
```suggestion
	viper.Set("committed.local_cache.size_bytes", 24*1024*1024)
```
1. Remove the comment - keeping it means we need to keep this file sync'ed up with defaults.go, and we won't do that.
1. We'll anyway probably never hit this in a small test, but if we do - it will slow down the API test.

arielshaqed reviewed (commented):
Leaving to @nopcoder to review properly.  Just preferences about one of the default values - which you should both feel free to ignore.

Overall this is really cool and exciting - looking forward to a faster api.test!!1!

nopcoder commented on pkg/pyramid/eviction.go:
```diff
@@ -25,8 +25,22 @@ const (
+	var numCountersToUse int64 = numCounters
+	if capacity < 100*1024*1024 { // If capacity < 100MB
+		// Use a more reasonable ratio: ~100 counters per 1MB of capacity
+		numCountersToUse = capacity / (1024 * 10) // ~100 counters per 10KB
+		if numCountersToUse < 1000 {
+			numCountersToUse = 1000 // Minimum threshold
```
set constants for the numbers with meaningful names, our lint rules don't like magic numbers.

nopcoder reviewed (changes requested):
looks good - thanks, left a comment related to the lint errors.

highlyavailable commented:
Thanks for the reviews! I've addressed both comments:

  1. @arielshaqed  Removed the comments from serve_test.go and updated the cache size to 24MB 
  2. @nopcoder  Added named constants in `eviction.go` to replace "magic numbers": 
    - smallCacheThreshold = 100MB
    - bytesPerCounter = 10KB per counter
    - minCounters = 1000

nopcoder reviewed (changes requested):
thanks for the quick response. <del>can you use gofmt and I'll merge it.</del> I'll merge and fmt after - thanks again!

nopcoder reviewed (approved):
LGTM!
